### ISSUE TITLE
[1.12.2] Client gets stuck when loading moon dimension

### ISSUE THREAD

**Issue by Zi0P4tch0** [micdoodle8/Galacticraft#3157]
REVIEW THE GUIDELINES (LINK AT TOP OF THIS PAGE)

Minecraft version: **1.12.2**
Galacticraft version: **4.0.1.132**
Forge version: **14.23.2.2611**

Using JDK 8 javaw with the following arguments:
-Xms4096m -Xmx4096m -XX:+UseParallelGC -XX:ParallelGCThreads=3 -XX:MaxGCPauseMillis=3

Single player (SSP) or Multiplayer (SMP)? **SMP**

Galacticraft add-ons and other mods installed? **Yes (ExtraPlanets and others)**

Hi there, I'm playing on a custom modpack which has both Galacticraft and ExtraPlanets.
I've tried disabling mods like BetterFPS and FoamFix, but the issue is still there.

Whenever I try to travel to the moon - via rocket - the client gets stuck at the loading screen (see attached picture)
GC debug messages are on.

During the startup I get this message on the client:

```
[07:51:48] [Netty Client IO #1/ERROR] [Galacticraft]: Galacticraft packet length problem for packet type C_SEND_OVERWORLD_IMAGE
[07:51:48] [Netty Client IO #1/ERROR] [Galacticraft]: Galacticraft packet length problem for packet type C_SEND_OVERWORLD_IMAGE
```

I don't know whether it's relevant to the issue or not.

![capture](https://user-images.githubusercontent.com/3139724/35904835-8f579b7e-0bdc-11e8-969d-d649424d1b37.PNG)

**SERVER LOG**

https://pastebin.com/miwPuthN

**CLIENT LOG**

https://pastebin.com/nJdvuMmt

**Comment by radfast** [comment]
Thanks for the careful report.

This **looks like** the cause is in Galacticraft itself.  We will investigate as a priority, but it could be 14 hours before there is any fix.

You can maybe temporarily fix the problem by deleting (or renaming) the folder ```world\galacticraft\overworldMap``` on your server - the server won't send these packets if that folder is empty.  But a fast server will be able to recreate that folder for itself in about 35 minutes, so you'll need to keep deleting it if you want to run the server today....

**Comment by Zi0P4tch0** [comment]
You're very welcome - I just checked mods on both clients and server, they match.
Thanks!

**Comment by roboderpy** [comment]
@radfast typing `/dimensiontp` is suffocating the executor of the command. If admin is not fast enough, he will die `<admin> suffocated in a wall`

**Comment by radfast** [comment]
@Zi0P4tch0  I can't reproduce this problem in testing.

Launching rockets from the Overworld is working 100% OK on my 1.12.2 test server.  Tested using Sponge and also tested using plain Forge (build 2586).

My test server has a small modpack including AE2, IC2, Mekanism, TC and Thermal Expansion only, so I don't have to wait too long for loading.

-----------------

Your issue is client-side crash here.  (The server log reads ```Zi0P4tch0 lost connection: Timed out``` which is typical for a client-side crash.)

I don't know what to think when there is nothing showing in the client log.

This must be an inter-mod conflict, not a bug in Galacticraft.  Please start by removing **Extra Planets** (and any other Galacticraft addons) from your mods folder.

If that does not fix it already, it is **extremely hard** for me to troubleshoot a 30+ mod list.  Please can you test a smaller group of mods, and smaller again (etc. - you can use a **binary search** system for efficiency) until you find the **1 mod** which is causing this.  So ... I hope you can find ... Galacticraft + X mod crashes on rocket launch but Galacticraft alone does not crash.

If you can tell me what X mod is there, then I can load it together with Galacticraft inside a debug environment and really figure out what is going on.

**Comment by radfast** [comment]
@roboderpy noted, added to my list of minor issues.  <admin> needs to be in Creative or Spectator mode for this one I guess....

**Comment by roboderpy** [comment]
Client doesn't seem to crash but waits for server for a long time. This **is** the original Minecraft issue (client/server just "ignore" that client waits for server to newly generate terrain and client timeouts itself from the server and server timeouts the client). You can reproduce this issue even with other (slow) dimension mods! Twilight Forest for example (entering into portal on new world would cause server think for some time and timeout the player both clientside and serverside)

**Comment by roboderpy** [comment]
I don't say it is "basically" slow, but without hacks in its code. Overworld in Minecraft generates quickly because it probably use some dirty methods and hacks. Again, this is not anyone's fault that terrain generates not fast enough 😕

**Comment by radfast** [comment]
>  If only could MC Forge API make something to allow modders to use async chunk gen provided by Mojang with threads...

You can probably code that, a smart guy like you.  Main thing to watch is that you'd need to put **all** of the async chunk gen into the same thread, because of the way Minecraft map generation over and over re-uses the same ```int[]``` objects from a small pool.  Or do like our Galacticraft background mapping function (which runs in its own thread) with a copy of the ```int[]``` pool, but this needs a bit of hacking with the classtransformer to make it work.

**Comment by radfast** [comment]
> Overworld in Minecraft generates quickly because it probably use some dirty methods and hacks. 

Moon generates faster than Overworld because the worldgen is simple - no water, trees, or planets etc, and only a single biome.  Overworld has about 13 different worldgen layers.

**Comment by roboderpy** [comment]
Oh yea, Moon generation speed doesn't seem to be a problem for me. Maybe this issue is really just related to some code and not to world generation.

**Comment by radfast** [comment]
... which brings us back to here

>   This must be an inter-mod conflict, not a bug in Galacticraft. Please start by removing Extra Planets (and any other Galacticraft addons) from your mods folder.

>   If that does not fix it already, it is extremely hard for me to troubleshoot a 30+ mod list. Please can you test a smaller group of mods, and smaller again until you find the 1 mod which is causing this

**Comment by roboderpy** [comment]
Also about modlist - some of mods can hook generation events and do some heavy calculations on world generation, and thus, slow down GC Dimensions.

**Comment by radfast** [comment]
dude, you are still shooting from the hip, that's maybe the 5th time now you are making unsubstantiated statements.  Galacticraft traps that hook and prevents other mods worldgen from running on our planets, apart from a small whitelist.  (We don't want TC Slime Islands, IC2 Rubber Trees, etc, on the Moon.)

**Comment by Zi0P4tch0** [comment]
Hi gents, thanks for the infos.
I did the binary search, and the conflicting mod is **InGame Info XML** by Lunatrius.
I'm a bit disappointed as I was expecting a more convoluted mod to be the troublemaker!

**Comment by radfast** [comment]
You're a **star**, thank you for the search. 

I can now test Galacticraft in detail with **InGame Info XML** and see where the problem is.  Might take 1-2 days.

Generally it is actually the smaller / newer mods which cause problems like this.  Any problems between Galacticraft and the other "big" mods would already have been found and fixed because somebody would have reported them before...

For today, can you play the game OK without InGame Info?

**Comment by Zi0P4tch0** [comment]
It was a false positive, my apologies. 
The real culprit seems to be **JourneyMap**.

Both crashes might be related - it feels like a UI rendering issue when switching to the moon dimension, which in turn triggers a timeout (this is my personal assumption, I'm an dev too, but I have no idea how Minecraft does its rendering).

The crash tend to occur when the server is started and switching to the moon after then.

I will make more tests with those two mods disabled and let you know what I find out.

**Comment by radfast** [comment]
Yes please do make sure ... We don't want to have a reputation for incompatibility with JourneyMap - a widely used mod - unless there definitely is a problem.

If there is a problem, we will of course look at fixing it or working with the JourneyMap devs.

**Comment by Zi0P4tch0** [comment]
**Tests so far**:

4GB RAM with JourneyMap -> 100% timeout when traveling to the moon
6GB RAM with JourneyMap -> 100% timeout
6GB RAM with Xaero's Map -> once in a while timeout
8GB RAM with Xaero's Map -> **TESTING**

**Comment by radfast** [comment]
Good news, I've found and fixed the problem.  A new build 133 will be available later today.

**Comment by radfast** [comment]
Build 133 is now available from the download site.  Also has a new feature: **Hydraulic Platform**

Thanks for being the first to report this problem ... sorry it took a few days, the cause was obscure. :)
